go/kbfs/libkbfs/subscription_manager.go

```diff
@@ -315,18 +310,127 @@ func (sm *subscriptionManager) SubscribePath(ctx context.Context,
+const folderBranchPollingInterval = time.Second
+
+func (sm *subscriptionManager) cancelAndDeleteFolderBranchPollerLocked(
+	sid SubscriptionID) (deleted bool) {
+	if cancel, ok := sm.folderBranchPollerCancelers[sid]; ok {
+		cancel()
+		delete(sm.folderBranchPollerCancelers, sid)
+		return true
+	}
+	return false
+}
+
+func (sm *subscriptionManager) cancelAndDeleteFolderBranchPoller(
+	sid SubscriptionID) (deleted bool) {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	return sm.cancelAndDeleteFolderBranchPollerLocked(sid)
+}
+
+func (sm *subscriptionManager) pollOnFolderBranchForSubscribePathRequest(
+	ctx context.Context, req subscribePathRequest, parsedPath *parsedPath) {
+	ticker := time.NewTicker(folderBranchPollingInterval)
+	for {
+		select {
+		case <-ticker.C:
+			fb, err := parsedPath.getFolderBranch(ctx, sm.config)
+			if err != nil {
+				_ = sm.cancelAndDeleteFolderBranchPoller(req.sid)
+				return
+			}
+
+			if fb == (data.FolderBranch{}) {
+				continue
+			}
+
+			// We have a folderBranch now! Go ahead and complete the
+			// sbuscription, and send a notification too.
+
+			sm.lock.Lock()
+			defer sm.lock.Unlock()
+			// Check if we're done while holding the lock to protect
+			// against racing against unsubscribe.
+			select {
+			case <-ctx.Done():
+				// No need to call cancelAndDeleteFolderBranchPollerLocked here
+				// since we always cancel and delete at the same tiem and if
+				// it's canceled it must have been deleted too.
+				return
+			default:
+			}
+
+			err = sm.subscribePathWithFolderBranchLocked(req, parsedPath, fb)
+			if err != nil {
+				sm.log.CErrorf(ctx,
+					"subscribePathWithFolderBranchLocked sid=%s err=%v", req.sid, err)
+			}
+
+			sm.notifier.OnPathChange(
+				sm.clientID, []SubscriptionID{req.sid},
+				req.path, []keybase1.PathSubscriptionTopic{req.topic})
+
+			_ = sm.cancelAndDeleteFolderBranchPollerLocked(req.sid)
+			return
+		case <-ctx.Done():
+			_ = sm.cancelAndDeleteFolderBranchPoller(req.sid)
+			return
+		}
+	}
+}
+
+func (sm *subscriptionManager) subscribePathWithoutFolderBranchLocked(
+	req subscribePathRequest, parsedPath *parsedPath) {
+	ctx, cancel := context.WithCancel(context.Background())
+	sm.folderBranchPollerCancelers[req.sid] = cancel
+	go sm.pollOnFolderBranchForSubscribePathRequest(ctx, req, parsedPath)
+}
+
+// SubscribePath implements the SubscriptionManager interface.
+func (sm *subscriptionManager) SubscribePath(ctx context.Context,
+	sid SubscriptionID, path string, topic keybase1.PathSubscriptionTopic,
+	deduplicateInterval *time.Duration) error {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+
+	parsedPath, err := parsePath(userPath(path))
+	if err != nil {
+		return err
+	}
+	fb, err := parsedPath.getFolderBranch(ctx, sm.config)
+	if err != nil {
+		return err
+	}
+	req := subscribePathRequest{
+		sid:                 sid,
+		path:                path,
+		topic:               topic,
+		deduplicateInterval: deduplicateInterval,
+	}
+	if fb != (data.FolderBranch{}) {
```

Comment:
Seems like we don't need to take the lock until right before this `if`, is that right?  No need to hold it when parsing or talking to FBO.

Reply:
I intentionally put it at the beginning. (The other comment in SubscribeNonPath was for this). It's specifically to protect a race where unsubscribe happens immediately after subscribe, and we end up calling unsubscribePath before these happen, and that'd be a no-op. At the point we start the polling thing, unsubscribe has already returned and there won't be any more unsubscribe calls in the future for this subscriptionID, then we end up with a subscription that shouldn't be there.

Reply:
I guess the specific worry is that the `getFolderBranch()` call would take a non-zero amount of time in some cases?  Everything else should be pretty fast and the lock placement shouldn't matter at all, but I guess if the subscribe is actually creating the FBO (not sure if that can actually happen in production, since presumably the GUI would have done something to access the TLF before subscribing to it?).  I'm also slightly worried about a deadlock here if the FBO needs to call back into the subscriber somehow, maybe due to something we add in the future.  I guess I don't have a simple alternative suggestion though.

But maybe we could move the lock to just above the `getFolderBranch()` call in that case?  I'm just thinking that it will be clearer to have it protecting only the things that matter.  I guess I don't feel too strongly about it.